internal/unionstore/art/node.go

```diff
@@ -0,0 +1,741 @@
+func minimum(a *artAllocator, an artNode) artNode {
+	switch an.kind {
+	case typeLeaf:
+		return an
+	case typeNode4:
+		n4 := an.node4(a)
+		if !n4.inplaceLeaf.addr.isNull() {
+			return n4.inplaceLeaf
+		}
+		if !n4.children[0].addr.isNull() {
+			return minimum(a, n4.children[0])
+		}
+	case typeNode16:
+		n16 := an.node16(a)
+		if !n16.inplaceLeaf.addr.isNull() {
+			return n16.inplaceLeaf
+		}
+		if !n16.children[0].addr.isNull() {
+			return minimum(a, n16.children[0])
+		}
+	case typeNode48:
+		n48 := an.node48(a)
+		if !n48.inplaceLeaf.addr.isNull() {
+			return n48.inplaceLeaf
+		}
+
+		idx := uint8(0)
+		for n48.present[idx>>n48s]&(1<<uint8(idx%n48m)) == 0 {
+			idx++
+		}
+		if !n48.children[n48.keys[idx]].addr.isNull() {
+			return minimum(a, n48.children[n48.keys[idx]])
+		}
+	case typeNode256:
+		n256 := an.node256(a)
+		if !n256.inplaceLeaf.addr.isNull() {
+			return n256.inplaceLeaf
+		} else {
+			for idx := 0; idx < 256; idx++ {
+				if !n256.children[idx].addr.isNull() {
+					return minimum(a, n256.children[idx])
+				}
+			}
+		}
+	}
+	return nullArtNode
+}
+
+// Find the minimum leaf under an artNode
+func maximum(a *artAllocator, an artNode) artNode {
+	switch an.kind {
+	case typeLeaf:
+		return an
+	case typeNode4:
+		n4 := an.node4(a)
+		if n4.nodeNum > 0 {
+			return maximum(a, n4.children[n4.nodeNum-1])
+		}
+		if !n4.inplaceLeaf.addr.isNull() {
+			return n4.inplaceLeaf
+		}
+	case typeNode16:
+		n16 := an.node16(a)
+		if n16.nodeNum > 0 {
+			return maximum(a, n16.children[n16.nodeNum-1])
+		}
+		if !n16.inplaceLeaf.addr.isNull() {
+			return n16.inplaceLeaf
+		}
+	case typeNode48:
+		n48 := an.node48(a)
+
+		idx := n48.prevPresentIdx(node256cap - 1)
+		if idx >= 0 {
+			return maximum(a, n48.children[n48.keys[idx]])
+		}
+		if !n48.inplaceLeaf.addr.isNull() {
+			return n48.inplaceLeaf
+		}
+	case typeNode256:
+		n256 := an.node256(a)
+		idx := n256.prevPresentIdx(node256cap - 1)
+		if idx >= 0 {
+			return maximum(a, n256.children[idx])
+		}
+		if !n256.inplaceLeaf.addr.isNull() {
+			return n256.inplaceLeaf
+		}
+	}
+	return nullArtNode
+}
+
+func (an artNode) findChild(a *artAllocator, c byte, valid bool) (int, artNode) {
+	node := an.node(a)
+	if !valid {
+		return -1, node.inplaceLeaf
+	}
+	switch an.kind {
+	case typeNode4:
+		return an._findChild4(a, c)
+	case typeNode16:
+		return an._findChild16(a, c)
+	case typeNode48:
+		return an._findChild48(a, c)
+	case typeNode256:
+		return an._findChild256(a, c)
+	}
+	return -2, nullArtNode
+}
+
+func (an artNode) _findChild4(a *artAllocator, c byte) (int, artNode) {
+	n4 := an.node4(a)
+	for idx := 0; idx < int(n4.nodeNum); idx++ {
+		if n4.keys[idx] == c {
+			return idx, n4.children[idx]
+		}
+	}
+	return -2, nullArtNode
+}
+
+func (an artNode) _findChild16(a *artAllocator, c byte) (int, artNode) {
+	n16 := an.node16(a)
+	for idx := 0; idx < int(n16.nodeNum); idx++ {
+		if n16.keys[idx] == c {
+			return idx, n16.children[idx]
+		}
+	}
```

Comment:
IIRC the paper uses SIMD or binary search here. We may benchmark the difference between the linear search and a binary search